evennia/plugin.py

```diff
@@ -0,0 +1,217 @@
+        return f"{self.__class__.__module__}.{self.__class__.__name__}"
+
+    def load_priority(self) -> int:
+        """
+        The order that plugin's load hooks will be called in. the lower the number, the sooner
+        it will be called.
+
+        Returns:
+            number
+        """
+        return 0
+
+    def init_settings(self, settings, plugins):
+        """
+        First hook to be called. Override this to make a plugin alter the Evennia settings.py
+        during the boot-up process.
+
+        WARNING: This is called by the launcher and possibly in other places, and the run state
+            of Evennia is not guaranteed. ONLY use this method to alter the settings.
+
+        Args:
+            settings (settings.py): The module reference to settings.py.
+            plugins (dict of name->EvPlugin classes): Allows plugins to know what
+                plugins will be loading.
+
+        Returns:
+            None
+        """
+
+    def at_plugin_setup(self):
+        """
+        First phase of plugin setup at the end of evennia._init()
+        """
+
+    def at_plugin_collaborate(self):
+        """
+        Second phase of plugin setup during evennia._init(). At this point, all plugins
+        are available in the evennia.PLUGINS dictionary.
+
+        Use this method for any plugin setup that involves plugins collaborating with each other.
+        """
+
+    def at_plugin_finalize(self):
+        """
+        Third and final phase of plugin setup during evennia._init(). Any last steps? Do them here.
+        """
+
+    def at_server_start(self):
+        """
+        Runs in the same order as load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+    def at_server_stop(self):
+        """
+        Runs in the reverse order of load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+    def at_server_reload_start(self):
+        """
+       Runs in the same order as load_priority, but before the 'mygame' hooks.
+
+       see at_server_startstop module documentation.
+       """
+
+    def at_server_reload_stop(self):
+        """
+        Runs in the reverse order of load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+    def at_server_cold_start(self):
+        """
+       Runs in the same order as load_priority, but before the 'mygame' hooks.
+
+       see at_server_startstop module documentation.
+       """
+
+    def at_server_cold_stop(self):
+        """
+        Runs in the reverse order of load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+
+class PluginManager:
+    """
+    Basic class used for maintaining an index of Plugins installed.
+    """
+
+    def __init__(self):
+        self.plugins = dict()
+        self.plugins_sorted = list()
```

Comment:
This is fine, but a nitpick is that it's faster and more explicit to assign `{}` and `[]` here rather than calling the functions. Also, it looks like `.reverse = None` goes poorly with the othe defaults; it would lead to a traceback when no plugins were yet added.